index/index.go

```diff
@@ -642,15 +601,56 @@ func newIndexReader(b ByteSlice, c io.Closer) (*indexReader, error) {
 	}
 	var err error
 
-	r.labels, err = r.readOffsetTable(r.toc.labelIndicesTable)
+	err = r.readOffsetTable(r.toc.labelIndicesTable, func(key []string, off uint32) error {
+		if len(key) != 1 {
+			return errors.Errorf("unexpected key length %d", len(key))
+		}
+		r.labels[key[0]] = off
+		return nil
+	})
 	if err != nil {
 		return nil, errors.Wrap(err, "read label index table")
 	}
-	r.postings, err = r.readOffsetTable(r.toc.postingsTable)
-	return r, errors.Wrap(err, "read postings table")
+	err = r.readOffsetTable(r.toc.postingsTable, func(key []string, off uint32) error {
+		if len(key) != 2 {
+			return errors.Errorf("unexpected key length %d", len(key))
+		}
+		r.postings[labels.Label{Name: key[0], Value: key[1]}] = off
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "read postings table")
+	}
+
+	r.dec = &DecoderV1{symbols: r.symbols}
+
+	return r, nil
+}
+
+// Range marks a byte range.
+type Range struct {
+	Start, End int64
+}
+
+// PostingsRanges returns a new map of byte range in the underlying index file
+// for all postings lists.
+func (r *Reader) PostingsRanges() (map[labels.Label]Range, error) {
```

Comment:
This is a weird addition, is it being used in thanos?

Reply:
Yea, I agree – mostly weird because it is _just_ for postings and nothing else? Or weird as data to expose in general.

For context, if we were to redo the format (or do a v2) I'd probably strictly move all references (except for symbols) to be of shape `(offset,length,crc32)` rather than just `offset`. It's really  painful to not know how much you have to fetch from `offset` – and that's not just true for object storages but even for a `File.Read` based implementation.

Reply:
+1 to include length! I thought about the exact same thing several times now!